README.md
# ChatGPT VSCode Extension

This is a Visual Studio Code extension for ChatGPT that was built using only ChatGPT and 5 beers.
The `README.md` file for this extension was generated by ChatGPT.

## Install the plugin from the [VSCode MarketPlace](https://marketplace.visualstudio.com/items?itemName=JayBarnes.chatgpt-vscode-plugin)

## Features
#### Commands:
- `ChatGPT: Query` (will provide a prompt for you to ask a question)
- `ChatGPT: Add tests` (generates tests based on file/text selection)
- `ChatGPT: Why is my code broken?` (analyses your code to highlight any logic/syntax errors)
- `ChatGPT: Explain code`
- `ChatGPT: Refactor`
- `ChatGPT: Reset token` (clears access token if expired or having issues)
- `ChatGPT: Reset thread` (resets your chat history)

*Everything except Reset Token and Query are available from the context menu when right-clicking in the editor.*


![image](https://user-images.githubusercontent.com/38425102/206071229-f017247e-831b-4e42-8c1a-914851da392f.png)

## Usage

To use the extension:
**This package requires a valid session token from ChatGPT to access it's unofficial REST API.**
Open the VS Code Command Palette and Type `ChatGPT: Reset token`, this will prompt you for your session token.

To get a session token:

1. Go to https://chat.openai.com/chat and log in or sign up.
2. Open dev tools.
3. Open `Application` > `Cookies` (`Storage` > `Cookies` on FireFox)
   
 ![image](https://user-images.githubusercontent.com/38425102/205900045-185c2c41-b4ff-408c-9da6-bbb606ac39c6.png)
   
4. Copy the value for `__Secure-next-auth.session-token` and enter it into the prompt from `ChatGPT: Reset token`

*Once you're logged in, you can ask ChatGPT any question and supply source code from your current file/selection.*

## Support
If you need help using this extension, please open an issue on the GitHub repository for this extension.

## Credits
- [ChatGPT](https://chat.openai.com/chat) - The large language model trained by OpenAI that was used to generate this README file
- [chatgpt-api](https://github.com/transitive-bullshit/chatgpt-api/) - The NPM package used to query ChatGPT
- [mpociot's extension](https://github.com/mpociot/chatgpt-vscode) - Inspiration for the project and the original webview panel
- [Gencay's extension](https://github.com/gencay/vscode-chatgpt) - Ported version of Gencay's webview panel.
- [Yeoman](https://yeoman.io/) - The code generator used to scaffold the extension project
- [VS Code Extension Generator](https://github.com/Microsoft/vscode-generator-code) - The Yeoman generator for creating VS Code extensions


CHANGELOG.md
# Change Log

All notable changes to the "chatgpt-vscode-plugin" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/test/chatPanel.ts
import * as vscode from 'vscode';
import { ChatGPTAPI } from '../chatgpt';

export class ChatPanel {
    public static currentPanel: ChatPanel | undefined;
    public static readonly viewType = 'chatgpt-vscode-plugin.chatView';
    private readonly _panel: vscode.WebviewPanel;
    private readonly _extensionUri: vscode.Uri;
    private _disposables: vscode.Disposable[] = [];
    private _chatApi: ChatGPTAPI;

    public constructor(panel: vscode.WebviewPanel, extensionUri: vscode.Uri) {
        this._panel = panel;
        this._extensionUri = extensionUri;
        this._panel.onDidDispose(() => this.dispose(), null, this._disposables);
        this._panel.webview.html = this.getHtml(this._panel.webview);
        this._chatApi = new ChatGPTAPI({ headless: false });
        this._panel.webview.onDidReceiveMessage(
            message => {
                switch (message.command) {
                    case 'alert':
                        vscode.window.showErrorMessage(message.text);
                        return;
                }
            }
        );
    }
    async search(prompt: any) {
        const isSignedIn = await this._chatApi.getIsSignedIn();
        if (!isSignedIn) {
            await this._chatApi.init();
            // await new ChatGPTAPI({ headless: false }).init();
        }

        const surroundingText = vscode.window.activeTextEditor?.document.getText();
        const selection = vscode.window.activeTextEditor?.selection;
        const selectedText = vscode.window.activeTextEditor?.document.getText(selection);
        let searchPrompt = '';

        if (selection && selectedText) {
            searchPrompt = `${selectedText}

${prompt}`;
        } else {
            searchPrompt = `This is the ${vscode.window.activeTextEditor?.document.languageId} file I'm working on:
${surroundingText}

${prompt}`;
        }

        const response = await this._chatApi.sendMessage(searchPrompt);
        this._panel.webview.postMessage({ type: 'addResponse', value: response });
    }

    getHtml(webview: vscode.Webview) {
        const scriptUri = webview.asWebviewUri(
            vscode.Uri.joinPath(this._extensionUri, 'media', 'lib.js')
        );

        return `<!DOCTYPE html>
			<html lang="en">
			<head>
				<meta charset="UTF-8">
				<meta name="viewport" content="width=device-width, initial-scale=1.0">
				<script src="  https://unpkg.com/showdown/dist/showdown.min.js"></script>
				<script src="https://cdn.tailwindcss.com/tailwind.min.js"></script>
				<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@400;700&display=swap" rel="stylesheet">
				<link href="https://cdn.tailwindcss.com/tailwind.min.css" rel="stylesheet">
				<style>
					:root {
						--vscode-editor-font-family: 'Roboto', sans-serif;
						--vscode-font-family: 'Roboto', sans-serif;
					}
	
					body {
						font-family: 'Roboto', sans-serif;
					}
				</style>
				<title>Chat with GPT-3</title>
			</head>
			<body>
				<div class="bg-gray-800 h-screen w-screen flex flex-col">
					<div class="p-4 h-full flex-grow flex flex-col">
						<div class="flex-grow overflow-auto">
							<div id="conversation" class="px-4 py-2 bg-gray-800 rounded-lg"></div>
						</div>
						<form onsubmit="sendMessage(); return false;" class="bg-gray-700 px-4 py-2 rounded-lg flex items-center">
							<input id="message" class="flex-grow px-2 py-1 rounded-lg" placeholder="Enter message...">
							<button type="submit" class="px-2 py-1 rounded-lg bg-gray-800 text-gray-100">
								Send
							</button>
						</form>
					</div>
				</div>
				<script src="${scriptUri}"></script>
			</body>
		</html>`;
    }

    public static createOrShow(extensionUri: vscode.Uri) {
        const column = vscode.window.activeTextEditor
            ? vscode.window.activeTextEditor.viewColumn
            : undefined;

        // If we already have a panel, show it.
        if (ChatPanel.currentPanel) {
            ChatPanel.currentPanel._panel.reveal(column);
            return;
        }

        // Otherwise, create a new panel.
        const panel = vscode.window.createWebviewPanel(
            ChatPanel.viewType,
            'ChatGPT',
            column || vscode.ViewColumn.One,
            {
                // Enable javascript in the webview
                enableScripts: true,

                // And restrict the webview to only loading content from our extension's `media` directory.
                localResourceRoots: [vscode.Uri.joinPath(extensionUri, 'media')],
            }
        );

        ChatPanel.currentPanel = new ChatPanel(panel, extensionUri);
    }

    public dispose() {
        ChatPanel.currentPanel = undefined;

        // Clean up our resources
        this._panel.dispose();

        while (this._disposables.length) {
            const x = this._disposables.pop();
            if (x) {
                x.dispose();
            }
        }
    }
}

media/lib.js
(function () {
    console.log('lib loaded');
    const vscode = acquireVsCodeApi();

    let response = '';

    // Handle messages sent from the extension to the webview
    window.addEventListener('message', event => {
        const message = event.data;
        switch (message.type) {
            case 'addResponse': {
                response = message.value;
                setResponse();
                break;
            }
            case 'clearResponse': {
                response = '';
                break;
            }
        }
    });

    function setResponse() {
        const converter = new showdown.Converter();
        const html = converter.makeHtml(response);
        document.getElementById('response').innerHTML = html;

        const preCodeBlocks = document.querySelectorAll('pre code');
        preCodeBlocks.forEach(block => {
            block.classList.add('p-4', 'my-4', 'block');
        });

        const codeBlocks = document.querySelectorAll('code');
        codeBlocks.forEach(block => {
            // Check if innertext starts with "Copy code"
            if (block.innerText.startsWith('Copy code')) {
                block.innerText = block.innerText.replace('Copy code', '');
            }

            block.classList.add(
                'p-1',
                'inline-flex',
                'max-w-full',
                'overflow-hidden',
                'border',
                'rounded-sm',
                'cursor-pointer'
            );

            block.addEventListener('click', function (e) {
                e.preventDefault();
                vscode.postMessage({
                    type: 'codeSelected',
                    value: this.innerText
                });
            });
        });
    }

    document.getElementById('prompt-input').addEventListener('keyup', function (e) {
        if (e.keyCode === 13) {
            vscode.postMessage({
                type: 'prompt',
                value: this.value
            });
        }
    });
})();

media/main.js
// @ts-nocheck

// This script will be run within the webview itself
// It cannot access the main VS Code APIs directly.
(function () {
    const vscode = acquireVsCodeApi();
    const list = document.getElementById("qa-list");

    // Handle messages sent from the extension to the webview
    window.addEventListener("message", (event) => {
        const message = event.data;

        switch (message.type) {
            case "addQuestion":
                const html = message.code != null
                    ? marked.parseInline(message.value + "<br /> <br /><pre class='overflow-auto'><code>```" + message.code + "```</code></pre>")
                    : message.value;

                list.innerHTML +=
                    `<div class="p-4 self-end mb-4">
                        <p class="font-bold mb-5 flex">
                            <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="w-5 h-5 mr-2"><path stroke-linecap="round" stroke-linejoin="round" d="M17.982 18.725A7.488 7.488 0 0012 15.75a7.488 7.488 0 00-5.982 2.975m11.963 0a9 9 0 10-11.963 0m11.963 0A8.966 8.966 0 0112 21a8.966 8.966 0 01-5.982-2.275M15 9.75a3 3 0 11-6 0 3 3 0 016 0z" /></svg>
                            You
                        </p>
                        <div>${html}</div>
                    </div>`;

                document.getElementById("in-progress")?.classList?.remove("hidden");
                list.scrollTo(0, list.scrollHeight);
                break;
            case "addResponse":
                document.getElementById("in-progress")?.classList?.add("hidden");

                list.innerHTML +=
                    `<div class="p-4 self-end mb-4 pb-8">
                        <p class="font-bold mb-5 flex">
                            <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="w-5 h-5 mr-2"><path stroke-linecap="round" stroke-linejoin="round" d="M8.25 3v1.5M4.5 8.25H3m18 0h-1.5M4.5 12H3m18 0h-1.5m-15 3.75H3m18 0h-1.5M8.25 19.5V21M12 3v1.5m0 15V21m3.75-18v1.5m0 15V21m-9-1.5h10.5a2.25 2.25 0 002.25-2.25V6.75a2.25 2.25 0 00-2.25-2.25H6.75A2.25 2.25 0 004.5 6.75v10.5a2.25 2.25 0 002.25 2.25zm.75-12h9v9h-9v-9z" /></svg>
                            ChatGPT
                        </p>
                        <div>${marked.parse(message.value)}</div>
                    </div>`;

                list.scrollTo(0, list.scrollHeight);
                break;
            default:
                break;
        }
    });

    let submitHandler = function (e) {
        e.preventDefault();
        e.stopPropagation();
        const input = document.getElementById("question-input");
        if (input.value?.length > 0) {
            vscode.postMessage({
                type: "askChatGPT",
                value: input.value,
            });

            input.value = "";
        }
    };

    document.getElementById("clear-button")?.addEventListener("click", () => {
        list.innerHTML = "";
        vscode.postMessage({ type: "clearChat", });
    });
    document.getElementById("ask-button")?.addEventListener("click", submitHandler);
    document.getElementById("question-input")?.addEventListener("keydown", function (e) {
        console.log(e.key);
        if (e.key === "Enter" && !e.shiftKey) {
            submitHandler(e);
        }
    });
})();


src/MyWebView.ts
import * as vscode from 'vscode';

export class MyWebview {
    private readonly panel: vscode.WebviewPanel;

    constructor() {
        // Create the Webview panel
        this.panel = vscode.window.createWebviewPanel(
            'webview', // Identifies the type of the webview. Used internally
            'Webview', // Title of the panel displayed to the user
            vscode.ViewColumn.One, // Editor column to show the new webview panel in.
            {
                enableScripts: true, // Allow scripts in the webview
            }
        );

        // Set the HTML content of the Webview
        this.panel.webview.html = `
            <!DOCTYPE html>
            <html>
            <head>
                <meta charset="UTF-8">
                <meta http-equiv="Content-Security-Policy" content="default-src 'none'; img-src https:; script-src 'unsafe-inline'; style-src vscode-resource: 'unsafe-inline';">
                <meta name="viewport" content="width=device-width, initial-scale=1.0">
                <title>My Webview</title>
            </head>
            <body>
                <h1>Hello from the Webview!</h1>
            </body>
            </html>
        `;
    }
}


src/extension.ts
import * as vscode from 'vscode';
import ChatGptViewProvider from './chatgpt-view-provider';

export async function activate(context: vscode.ExtensionContext) {
	const chatViewProvider = new ChatGptViewProvider(context);

	context.subscriptions.push(
		vscode.commands.registerCommand('chatgpt-vscode-plugin.askGPT', askChatGPT),
		vscode.commands.registerCommand('chatgpt-vscode-plugin.whyBroken', askGPTWhyBroken),
		vscode.commands.registerCommand('chatgpt-vscode-plugin.explainPls', askGPTToExplain),
		vscode.commands.registerCommand('chatgpt-vscode-plugin.refactor', askGPTToRefactor),
		vscode.commands.registerCommand('chatgpt-vscode-plugin.addTests', askGPTToAddTests),
		vscode.commands.registerCommand('chatgpt-vscode-plugin.resetToken', resetToken),
		vscode.commands.registerCommand('chatgpt-vscode-plugin.resetThread', resetThread),
		vscode.window.registerWebviewViewProvider("chatgpt-vscode-plugin.view", chatViewProvider, {
			webviewOptions: { retainContextWhenHidden: true }
		})
	);

	async function askGPTToExplain() { await askChatGPT('Can you explain what this code does?'); }
	async function askGPTWhyBroken() { await askChatGPT('Why is this code broken?'); }
	async function askGPTToRefactor() { await askChatGPT('Can you refactor this code and explain what\'s changed?'); }
	async function askGPTToAddTests() { await askChatGPT('Can you add tests for this code?'); }

	async function resetToken() {
		await context.globalState.update('chatgpt-session-token', null);
		await chatViewProvider.setUpSessionToken();
		// await vscode.window.showInformationMessage("Token reset, you'll be prompted for it next to you next ask ChatGPT a question.");
	}

	async function resetThread() { await chatViewProvider.resetThread(); }

	async function askChatGPT(userInput?: string) {
		if (!userInput) {
			userInput = await vscode.window.showInputBox({ prompt: "Ask ChatGPT a question" }) || "";
		}

		let editor = vscode.window.activeTextEditor;

		if (editor) {
			const selectedCode = editor.document.getText(vscode.window.activeTextEditor?.selection);
			const entireFileContents = editor.document.getText();

			const code = selectedCode
				? selectedCode
				: `This is the ${editor.document.languageId} file I'm working on: \n\n${entireFileContents}`;

			chatViewProvider.sendApiRequest(userInput, code);
		}
	}
}

src/chatgpt/index.js
var k=Object.create;var w=Object.defineProperty;var v=Object.getOwnPropertyDescriptor;var C=Object.getOwnPropertyNames;var M=Object.getPrototypeOf,T=Object.prototype.hasOwnProperty;var D=(s,e)=>()=>(e||s((e={exports:{}}).exports,e),e.exports);var B=(s,e,r,a)=>{if(e&&typeof e=="object"||typeof e=="function")for(let t of C(e))!T.call(s,t)&&t!==r&&w(s,t,{get:()=>e[t],enumerable:!(a=v(e,t))||a.enumerable});return s};var L=(s,e,r)=>(r=s!=null?k(M(s)):{},B(e||!s||!s.__esModule?w(r,"default",{value:s,enumerable:!0}):r,s));var x=D((A,l)=>{"use strict";var E=(s,e)=>Math.floor(Math.random()*(e-s+1)+s),_=()=>{let s=new Error("Delay aborted");return s.name="AbortError",s},b=({clearTimeout:s,setTimeout:e,willResolve:r})=>(a,{value:t,signal:o}={})=>{if(o&&o.aborted)return Promise.reject(_());let i,n,m,g=s||clearTimeout,u=()=>{g(i),m(_())},P=()=>{o&&o.removeEventListener("abort",u)},p=new Promise((I,d)=>{n=()=>{P(),r?I(t):d(t)},m=d,i=(e||setTimeout)(n,a)});return o&&o.addEventListener("abort",u,{once:!0}),p.clear=()=>{g(i),i=null,n()},p},f=s=>{let e=b({...s,willResolve:!0});return e.reject=b({...s,willResolve:!1}),e.range=(r,a,t)=>e(E(r,a),t),e},c=f();c.createWithTimers=f;l.exports=c;l.exports.default=c});var h=L(x(),1);import $ from"html-to-md";import{chromium as U}from"playwright";var y=class{constructor(e={}){let{userDataDir:r="/tmp/chatgpt",chatUrl:a="https://chat.openai.com/",headless:t=!1,markdown:o=!0}=e;this._userDataDir=r,this._headless=!!t,this._chatUrl=a,this._markdown=!!o}async init(e={}){let{auth:r="eager"}=e;this._browser&&await this.close(),this._browser=await U.launchPersistentContext(this._userDataDir,{headless:this._headless}),this._page=await this._browser.newPage(),await this._page.goto(this._chatUrl);do{let a='[data-headlessui-state="open"]';if(!await this._page.$(a))break;try{await this._page.click(`${a} button:last-child`,{timeout:1e3})}catch{break}}while(!0);if(r==="blocking")do{if(await this.getIsSignedIn())break;console.log("Please sign in to ChatGPT using the Chromium browser window and dismiss the welcome modal..."),await(0,h.default)(1e3)}while(!0);return this._page}async getIsSignedIn(){try{return!!await this._getInputBox()}catch{return!1}}async getLastMessage(){let e=await this.getMessages();return e?e[e.length-1]:null}async getPrompts(){let e=await this._page.$$('[class*="ConversationItem__Message"]:has([class*="ConversationItem__ActionButtons"]):has([class*="ConversationItem__Role"] [class*="Avatar__Wrapper"])');return Promise.all(e.map(r=>r.innerText()))}async getMessages(){let e=await this._page.$$('[class*="ConversationItem__Message"]:has([class*="ConversationItem__ActionButtons"]):not(:has([class*="ConversationItem__Role"] [class*="Avatar__Wrapper"]))');return this._markdown?(await Promise.all(e.map(t=>t.innerHTML()))).map(t=>(t=t.replace("Copy code</button>","</button>"),$(t,{ignoreTags:["button","svg","style","form","noscript","script","meta","head"],skipTags:["button","svg"]}))):await Promise.all(e.map(a=>a.innerText()))}async sendMessage(e){let r=await this._getInputBox();if(!r)throw new Error("not signed in");let a=await this.getLastMessage();await r.click({force:!0}),await r.fill(e,{force:!0}),await r.press("Enter");do{await(0,h.default)(1e3);let t=await this.getLastMessage();if(t&&(a==null?void 0:a.toLowerCase())!==(t==null?void 0:t.toLowerCase()))return t}while(!0)}async close(){await this._browser.close(),this._page=null,this._browser=null}async _getInputBox(){return this._page.$('div[class*="PromptTextarea__TextareaWrapper"] textarea')}};export{y as ChatGPTAPI};
//# sourceMappingURL=index.js.map

src/chatgpt/index.d.ts
import { ChromiumBrowserContext, Page } from 'playwright';

declare class ChatGPTAPI {
    protected _userDataDir: string;
    protected _headless: boolean;
    protected _markdown: boolean;
    protected _chatUrl: string;
    protected _browser: ChromiumBrowserContext;
    protected _page: Page;
    /**
     * @param opts.userDataDir — Path to a directory for storing persistent chromium session data
     * @param opts.chatUrl — OpenAI chat URL
     * @param opts.headless - Whether or not to use headless mode
     * @param opts.markdown — Whether or not to parse chat messages as markdown
     */
    constructor(opts?: {
        /** @defaultValue `'/tmp/chatgpt'` **/
        userDataDir?: string;
        /** @defaultValue `'https://chat.openai.com/'` **/
        chatUrl?: string;
        /** @defaultValue `false` **/
        headless?: boolean;
        /** @defaultValue `true` **/
        markdown?: boolean;
    });
    init(opts?: {
        auth?: 'blocking' | 'eager';
    }): Promise<Page>;
    getIsSignedIn(): Promise<boolean>;
    getLastMessage(): Promise<string | null>;
    getPrompts(): Promise<string[]>;
    getMessages(): Promise<string[]>;
    sendMessage(message: string): Promise<string>;
    close(): Promise<void>;
    protected _getInputBox(): Promise<any>;
}

/**
 * https://chat.openapi.com/api/auth/session
 */
type SessionResult = {
    /**
     * Object of the current user
     */
    user: APIUser;
    /**
     * ISO date of the expiration date of the access token
     */
    expires: string;
    /**
     * The access token
     */
    accessToken: string;
};
type APIUser = {
    /**
     * ID of the user
     */
    id: string;
    /**
     * Name of the user
     */
    name: string;
    /**
     * Email of the user
     */
    email: string;
    /**
     * Image of the user
     */
    image: string;
    /**
     * Picture of the user
     */
    picture: string;
    /**
     * Groups the user is in
     */
    groups: string[] | [];
    /**
     * Features the user is in
     */
    features: string[] | [];
};
/**
 * https://chat.openapi.com/backend-api/models
 */
type ModelsResult = {
    /**
     * Array of models
     */
    models: APIModel[];
};
type APIModel = {
    /**
     * Name of the model
     */
    slug: string;
    /**
     * Max tokens of the model
     */
    max_tokens: number;
    /**
     * Whether or not the model is special
     */
    is_special: boolean;
};
/**
 * https://chat.openapi.com/backend-api/moderations
 */
type ModerationsJSONBody = {
    /**
     * Input for the moderation decision
     */
    input: string;
    /**
     * The model to use in the decision
     */
    model: AvailableModerationModels;
};
type AvailableModerationModels = 'text-moderation-playground';
/**
 * https://chat.openapi.com/backend-api/moderations
 */
type ModerationsJSONResult = {
    /**
     * Whether or not the input is flagged
     */
    flagged: boolean;
    /**
     * Whether or not the input is blocked
     */
    blocked: boolean;
    /**
     * The ID of the decision
     */
    moderation_id: string;
};
/**
 * https://chat.openapi.com/backend-api/conversation
 */
type ConversationJSONBody = {
    /**
     * The action to take
     */
    action: string;
    /**
     * The ID of the conversation
     */
    conversation_id?: string;
    /**
     * Prompts to provide
     */
    messages: APIPrompt[];
    /**
     * The model to use
     */
    model: string;
    /**
     * The parent message ID
     */
    parent_message_id: string;
};
type APIPrompt = {
    /**
     * The content of the prompt
     */
    content: APIPromptContent;
    /**
     * The ID of the prompt
     */
    id: string;
    /**
     * The role played in the prompt
     */
    role: APIPromptRole;
};
type APIPromptContent = {
    /**
     * The content type of the prompt
     */
    content_type: APIPromptContentType;
    /**
     * The parts to the prompt
     */
    parts: string[];
};
type APIPromptContentType = 'text';
type APIPromptRole = 'user';
/**
 * https://chat.openapi.com/backend-api/conversation/message_feedback
 */
type MessageFeedbackJSONBody = {
    /**
     * The ID of the conversation
     */
    conversation_id: string;
    /**
     * The message ID
     */
    message_id: string;
    /**
     * The rating
     */
    rating: APIMessageFeedbackRating;
    /**
     * Tags to give the rating
     */
    tags?: APIMessageFeedbackTags[];
    /**
     * The text to include
     */
    text?: string;
};
type APIMessageFeedbackTags = 'harmful' | 'false' | 'not-helpful';
type MessageFeedbackResult = {
    /**
     * The message ID
     */
    message_id: string;
    /**
     * The ID of the conversation
     */
    conversation_id: string;
    /**
     * The ID of the user
     */
    user_id: string;
    /**
     * The rating
     */
    rating: APIMessageFeedbackRating;
    /**
     * The text the server received, including tags
     */
    text?: string;
};
type APIMessageFeedbackRating = 'thumbsUp' | 'thumbsDown';

export { APIMessageFeedbackRating, APIMessageFeedbackTags, APIModel, APIPrompt, APIPromptContent, APIPromptContentType, APIPromptRole, APIUser, AvailableModerationModels, ChatGPTAPI, ConversationJSONBody, MessageFeedbackJSONBody, MessageFeedbackResult, ModelsResult, ModerationsJSONBody, ModerationsJSONResult, SessionResult };


src/chatgpt-view-provider.ts
import { ChatGPTAPI, ChatGPTConversation } from 'chatgpt';
import * as vscode from 'vscode';

export default class ChatGptViewProvider implements vscode.WebviewViewProvider {
    private webView?: vscode.WebviewView;
    private chatGptApi?: ChatGPTAPI;
    private chatGptConversaion?: ChatGPTConversation;
    private sessionToken?: string;
    private message?: any;

    constructor(private context: vscode.ExtensionContext) { }

    public resolveWebviewView(
        webviewView: vscode.WebviewView,
        _context: vscode.WebviewViewResolveContext,
        _token: vscode.CancellationToken,
    ) {
        this.webView = webviewView;
        webviewView.webview.options = {
            enableScripts: true,
            localResourceRoots: [this.context.extensionUri]
        };

        webviewView.webview.html = this.getHtml(webviewView.webview);

        webviewView.webview.onDidReceiveMessage(data => {
            if (data.type === 'askChatGPT') {
                this.sendApiRequest(data.value);
            }
            if (data.type === 'clearChat') {
                this.chatGptConversaion = this.chatGptApi?.getConversation();
            }
        });

        if (this.message !== null) {
            this.sendMessage(this.message);
            this.message = null;
        }
    }

    public async setUpSessionToken() {
        const sessionTokenName = 'chatgpt-session-token';
        this.sessionToken = await this.context.globalState.get(sessionTokenName) as string;

        if (!this.sessionToken) {
            const userSessionToken = await vscode.window.showInputBox({
                prompt: "Please enter your token (__Secure-next-auth.session-token), this can be retrieved using the guide on the README "
            });
            this.sessionToken = userSessionToken!;
            this.context.globalState.update(sessionTokenName, this.sessionToken);
        }
    }

    public async sendApiRequest(prompt: string, code?: string) {
        await this.setUpSessionToken();

        if (!this.chatGptApi || !this.chatGptConversaion) {
            try {
                this.chatGptApi = new ChatGPTAPI({ sessionToken: this.sessionToken as string });
                this.chatGptConversaion = this.chatGptApi?.getConversation();
            } catch (error: any) {
                vscode.window.showErrorMessage("Failed to connect to ChatGPT", error?.message);
                return;
            }
        }

        // Create question by adding prompt prefix to code, if provided
        const question = (code) ? `${prompt}: ${code}` : prompt;

        if (!this.webView) {
            await vscode.commands.executeCommand('chatgpt-vscode-plugin.view.focus');
        } else {
            this.webView?.show?.(true);
        }

        // Send question to ChatGPT and show response
        this.sendMessage({ type: 'addQuestion', value: prompt, code });
        try {
            await this.chatGptApi?.ensureAuth();
            const response = await this.chatGptConversaion.sendMessage(question, {
                timeoutMs: 2 * 60 * 1000
            });

            this.sendMessage({ type: 'addResponse', value: response });
        } catch (error: any) {
            await vscode.window.showErrorMessage("Error sending request to ChatGPT", error?.message);
            return;
        }
    }

    public sendMessage(message: any) {
        if (this.webView) {
            this.webView?.webview.postMessage(message);
        } else {
            this.message = message;
        }
    }

    private getHtml(webview: vscode.Webview) {

        const scriptUri = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'media', 'main.js'));
        const stylesMainUri = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'media', 'main.css'));

        return `<!DOCTYPE html>
			<html lang="en">
			<head>
				<meta charset="UTF-8">
				<meta name="viewport" content="width=device-width, initial-scale=1.0">
				<link href="${stylesMainUri}" rel="stylesheet">
				<script src="https://cdn.jsdelivr.net/npm/marked/marked.min.js"></script>
				<script src="https://cdn.tailwindcss.com"></script>
			</head>
			<body class="overflow-hidden">
				<div class="flex flex-col h-screen">
					<div class="flex-1 overflow-y-auto" id="qa-list"></div>
					<div id="in-progress" class="p-4 flex items-center hidden">
                        <div style="text-align: center;">
                            <div>Please wait while we handle your request ❤️</div>
                            <div class="loader"></div>
                            <div>Please note, ChatGPT facing scaling issues which will impact this extension</div>
                        </div>
					</div>
					<div class="p-4 flex items-center">
						<div class="flex-1">
							<textarea
								type="text"
								rows="2"
								class="border p-2 w-full"
								id="question-input"
								placeholder="Ask a question..."
							></textarea>
						</div>
						<button style="background: var(--vscode-button-background)" id="ask-button" class="p-2 ml-5">Ask</button>
						<button style="background: var(--vscode-button-background)" id="clear-button" class="p-2 ml-3">Clear</button>
					</div>
				</div>
				<script src="${scriptUri}"></script>
			</body>
			</html>`;
    }
}

src/extension copy.ts
import * as vscode from 'vscode';
import { ChatGPTAPI } from './chatgpt';

export function activate(context: vscode.ExtensionContext) {
	const provider = new GPTViewProvider(context.extensionUri);

	context.subscriptions.push(
		vscode.window.registerWebviewViewProvider(GPTViewProvider.viewType, provider)
	);

	context.subscriptions.push(
		vscode.commands.registerCommand('chatgpt-vscode-plugin.askGPT', () => {
			vscode.window.showInputBox({ prompt: 'What do you want to do?' }).then((value) => {
				provider.search(value);
			});
		})
	);
}

class GPTViewProvider implements vscode.WebviewViewProvider {
	public static readonly viewType = 'chatgpt-vscode-plugin.chatView';

	private _view?: vscode.WebviewView;
	private _chatGPTAPI = new ChatGPTAPI({ headless: false });

	constructor(private readonly _extensionUri: vscode.Uri) { }

	public resolveWebviewView(
		webviewView: vscode.WebviewView,
		context: vscode.WebviewViewResolveContext,
		_token: vscode.CancellationToken
	) {
		this._view = webviewView;
		this.configureWebview(webviewView.webview);
		webviewView.webview.html = this.getHtmlForWebview(webviewView.webview);
		webviewView.webview.onDidReceiveMessage(data => {
			switch (data.type) {
				case 'codeSelected': {
					const code = data.value.replace(/([^\\])(\$)([^{0-9])/g, "$1\\$$$3");
					vscode.window.activeTextEditor?.insertSnippet(new vscode.SnippetString(code));
					break;
				}
				case 'prompt': {
					this.search(data.value);
				}
			}
		});
	}

	public async search(prompt: any) {
		const isSignedIn = await this._chatGPTAPI.getIsSignedIn();
		if (!isSignedIn) {
			await this._chatGPTAPI.init();
			// await new ChatGPTAPI({ headless: false }).init();
		}

		const surroundingText = vscode.window.activeTextEditor?.document.getText();
		const selection = vscode.window.activeTextEditor?.selection;
		const selectedText = vscode.window.activeTextEditor?.document.getText(selection);
		let searchPrompt = '';

		if (selection && selectedText) {
			searchPrompt = `${selectedText}

${prompt}`;
		} else {
			searchPrompt = `This is the ${vscode.window.activeTextEditor?.document.languageId} file I'm working on:
${surroundingText}

${prompt}`;
		}

		const response = await this._chatGPTAPI.sendMessage(searchPrompt);
		if (this._view) {
			this._view.show?.(true);
			this._view.webview.postMessage({ type: 'addResponse', value: response });
		}
	}

	private configureWebview(webview: vscode.Webview) {
		webview.options = {
			enableScripts: true,
			localResourceRoots: [this._extensionUri],
		};
	}

	private getHtmlForWebview(webview: vscode.Webview) {
		const scriptUri = webview.asWebviewUri(
			vscode.Uri.joinPath(this._extensionUri, 'media', 'lib.js')
		);

		return `<!DOCTYPE html>
			<html lang="en">
			<head>
				<meta charset="UTF-8">
				<meta name="viewport" content="width=device-width, initial-scale=1.0">
				<script src="  https://unpkg.com/showdown/dist/showdown.min.js"></script>
				<script src="https://cdn.tailwindcss.com/tailwind.min.js"></script>
				<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@400;700&display=swap" rel="stylesheet">
				<link href="https://cdn.tailwindcss.com/tailwind.min.css" rel="stylesheet">
				<style>
					:root {
						--vscode-editor-font-family: 'Roboto', sans-serif;
						--vscode-font-family: 'Roboto', sans-serif;
					}
	
					body {
						font-family: 'Roboto', sans-serif;
					}
				</style>
				<title>Chat with GPT-3</title>
			</head>
			<body>
				<div class="bg-gray-800 h-screen w-screen flex flex-col">
					<div class="p-4 h-full flex-grow flex flex-col">
						<div class="flex-grow overflow-auto">
							<div id="conversation" class="px-4 py-2 bg-gray-800 rounded-lg"></div>
						</div>
						<form onsubmit="sendMessage(); return false;" class="bg-gray-700 px-4 py-2 rounded-lg flex items-center">
							<input id="message" class="flex-grow px-2 py-1 rounded-lg" placeholder="Enter message...">
							<button type="submit" class="px-2 py-1 rounded-lg bg-gray-800 text-gray-100">
								Send
							</button>
						</form>
					</div>
				</div>
				<script src="${scriptUri}"></script>
			</body>
		</html>`;
	}
}